lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,200 @@
+      break;
+    }
+  }
+  // Handle load instruction which is following the bitcast in the pattern
+  for (auto *U : OldInstIter->users()) {
+    if (auto *LI = dyn_cast<LoadInst>(U)) {
+      NewValue = Builder.CreateLoad(NewVecTy, NewValue);
+      OldInstIter = LI;
+      InstsToErase.push_back(OldInstIter);
+      break;
+    }
+  }
+  // Handle extractelement instruction which is following the load
+  for (auto *U : OldInstIter->users()) {
+    if (auto *EEI = dyn_cast<ExtractElementInst>(U)) {
+      VectorType *OldVecTy = getVectorType(Inst->getType());
+      uint64_t NumElemsInOldVec = OldVecTy->getElementCount().getValue();
+      uint64_t NumElemsInNewVec = NewVecTy->getElementCount().getValue();
+      uint64_t ElemIdx =
+          cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
+          (NumElemsInOldVec / NumElemsInNewVec);
+      NewValue = Builder.CreateExtractElement(NewValue, ElemIdx);
+      NewValue = Builder.CreateTrunc(NewValue, OldVecTy->getElementType());
+      OldInstIter = EEI;
+      InstsToErase.push_back(OldInstIter);
+      break;
+    }
+  }
+  OldInstIter->replaceAllUsesWith(NewValue);
+  return true;
+}
+
+class SPIRVLowerBitCastToNonStandardTypePass {
+public:
+  SPIRVLowerBitCastToNonStandardTypePass() {}
+
+  PreservedAnalyses
+  runLowerBitCastToNonStandardType(Function &F, FunctionAnalysisManager &FAM) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known. We assume that bad type won't be passed to a function as
+    // parameter, since it added by an optimization.
+    bool Changed = false;
+
+    std::vector<Instruction *> BCastsToNonStdVec;
+    std::vector<Instruction *> InstsToErase;
+    for (auto &BB : F)
+      for (auto &I : BB) {
+        auto *BC = dyn_cast<BitCastInst>(&I);
+        if (!BC)
+          continue;
+        VectorType *SrcVecTy = getVectorType(BC->getSrcTy());
+        if (SrcVecTy) {
+          uint64_t NumElemsInSrcVec = SrcVecTy->getElementCount().getValue();
+          if (!isValidVectorSize(NumElemsInSrcVec))
+            report_fatal_error("Unsupported vector type with the size of: " +
+                                   std::to_string(NumElemsInSrcVec),
+                               false);
+        }
+        VectorType *DestVecTy = getVectorType(BC->getDestTy());
+        if (DestVecTy) {
+          uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+          if (!isValidVectorSize(NumElemsInDestVec))
+            BCastsToNonStdVec.push_back(&I);
+        }
+      }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    for (auto I = InstsToErase.rbegin(); I != InstsToErase.rend(); ++I)
+      (*I)->eraseFromParent();
+
+    return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+};
+
+class SPIRVLowerBitCastToNonStandardTypeLegacy
+    : public FunctionPass,
+      public SPIRVLowerBitCastToNonStandardTypePass {
```

Comment:
```suggestion
    : public FunctionPass {
```

Why do you need to inherit from `SPIRVLowerBitCastToNonStandardTypePass` here?